src/coord/src/client.rs

```diff
@@ -481,6 +440,47 @@ impl SessionClient {
         Ok(SimpleExecuteResponse { results })
     }
 
+    /// Like [`SessionClient::simple_execute`], but for `TAIL` queries.
+    pub async fn simple_tail(
+        &mut self,
+        stmt: &str,
+    ) -> Result<impl Stream<Item = Result<Vec<Value>, CoordError>>, CoordError> {
+        let stmts = mz_sql::parse::parse(&stmt).map_err(|e| CoordError::Unstructured(e.into()))?;
+        if stmts.len() != 1 {
+            coord_bail!("only one statement is supported");
+        }
+        self.start_transaction(None).await?;
+        const EMPTY_PORTAL: &str = "";
+        self.declare(EMPTY_PORTAL.into(), stmts.into_element(), vec![])
+            .await?;
+        let desc = self
+            .session()
+            .get_portal_unverified(EMPTY_PORTAL)
+            .map(|portal| portal.desc.clone())
+            .expect("unnamed portal should be present");
+        if !desc.param_types.is_empty() {
+            return Err(CoordError::Unsupported("parameters"));
+        }
+        match self.execute(EMPTY_PORTAL.into()).await? {
+            ExecuteResponse::Tailing { mut rx } => Ok(stream! {
+                while let Some(res) = rx.recv().await {
+                    match res {
+                        PeekResponseUnary::Rows(rows) => {
+                            let mut datum_vec = mz_repr::DatumVec::new();
+                            for row in rows {
+                                let datums = datum_vec.borrow_with(&row);
+                                yield Ok(datums.iter().map(datum_to_json).collect());
+                            }
+                        }
+                        PeekResponseUnary::Error(e) => yield Err(CoordError::Unstructured(anyhow!(e))),
+                        PeekResponseUnary::Canceled => yield Err(CoordError::Unstructured(anyhow!("execution canceled"))),
+                    }
+                }
+            }),
+            _ => coord_bail!("statements of the executed type"),
```

Comment:
I've been angry at myself for using the pattern in simple_execute that's now copied here. We are returning an error because we don't know how to serialize the response, but the query actually ran, so we're lying to the users about if it happened or not. We're lucky today that we disallow ddls in txns because that has the side effect of preventing them here, but that's not a good foundation. We need to not error here and instead serialize all of these correctly, or otherwise prevent them before execution.

We can probably get to a point where both the WS and non WS interfaces use the same function to convert execute responses to http responses (maybe? perhaps we'd want to disallow tail responses in the non ws case? dunno).

Reply:
> I've been angry at myself for using the pattern in simple_execute that's now copied here. We are returning an error because we don't know how to serialize the response, but the query actually ran, so we're lying to the users about if it happened or not. We're lucky today that we disallow ddls in txns because that has the side effect of preventing them here, but that's not a good foundation. We need to not error here and instead serialize all of these correctly, or otherwise prevent them before execution.

Yeah I noticed that too! Would be great to fix.